Set the required version of why3find in the opam file

The instructions provided in the readme of the Creusot IDE VSCode extension currently don't work. The workaround is to use `opam install . --locked`, creating a switch just for Creusot. I'm also curious why that switch isn't `~/.creusot/_opam`, and if there's a risk of a clash.

```
[ERROR] Could not synchronize /Users/sacha/.opam/creusot/.opam-switch/sources/why3 from
        "git+ssh://git@gitlab.inria.fr/why3/why3#hyp_name":
        "/usr/bin/git fetch -q" exited with code 128
```
Ah I had missed that one

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -31,10 +31,7 @@ jobs:
         with:
           ocaml-compiler: 5
 
-      - run: opam pin add why3 https://gitlab.inria.fr/why3/why3.git#2c0f2992af85f82f3eda0f158dcf10e62e0db875
-      - run: opam pin add why3find https://git.frama-c.com/pub/why3find.git#eab37557d3e24e1913a3c4f44bc5528ef497c6c9
-
-      - run: opam install . --deps-only --with-test
+      - run: opam install . --deps-only --with-test --locked
 
       - run: opam exec -- dune build
 

--- a/README.md
+++ b/README.md
@@ -116,14 +116,14 @@ Available in the command palette (`Ctrl+P`):
 
 - To build and install the VS Code extension from source:
 
-```
-npx vsce package
-```
+    ```
+    npx vsce package
+    ```
 
-If esbuild complains about not being able to resolve "vscode-languageclient/node", try:
+    Then, in VS Code: Command Palette > Install from VSIX. Select the `creusot-ide-X.Y.Z.vsix` file created by the previous command.
 
-```
-npm install vscode-languageclient
-```
+    If esbuild complains about not being able to resolve "vscode-languageclient/node", try:
 
-Then, in VS Code: Command Palette > Install from VSIX. Select the `creusot-ide-X.Y.Z.vsix` file created by the previous command.
+    ```
+    npm install vscode-languageclient
+    ```

--- a/README.md
+++ b/README.md
@@ -63,6 +63,7 @@ At the moment, installing `creusot-lsp` in another switch or without opam altoge
     "rust-analyzer.check.overrideCommand": [
             "cargo",
             "creusot",
+            "--only=coma",
             "--",
             "--message-format=json"
         ]

--- a/creusot-lsp.opam.locked
+++ b/creusot-lsp.opam.locked
@@ -16,40 +16,36 @@ depends: [
   "base-threads" {= "base"}
   "base-unix" {= "base"}
   "bigstringaf" {= "0.10.0"}
-  "cmdliner" {= "1.3.0"}
+  "camlzip" {= "1.14"}
   "conf-autoconf" {= "0.2"}
   "conf-gmp" {= "5"}
-  "conf-libev" {= "4-13"}
   "conf-pkg-config" {= "4"}
+  "conf-zlib" {= "1"}
   "cppo" {= "1.8.0"}
   "csexp" {= "1.5.2"}
-  "dune" {= "3.20.2"}
-  "dune-build-info" {= "3.20.2"}
-  "dune-configurator" {= "3.20.2"}
-  "dune-private-libs" {= "3.20.2"}
-  "dune-site" {= "3.20.2"}
-  "dyn" {= "3.20.2"}
-  "fmt" {= "0.11.0"}
-  "gen" {= "1.1"}
-  "hmap" {= "0.8.1"}
+  "dune" {= "3.22.2"}
+  "dune-build-info" {= "3.22.2"}
+  "dune-configurator" {= "3.22.2"}
+  "dune-private-libs" {= "3.22.2"}
+  "dune-site" {= "3.22.2"}
+  "dyn" {= "3.22.2"}
+  "fs-io" {= "3.22.2"}
   "ISO8601" {= "0.2.6"}
   "jane-street-headers" {= "v0.17.0"}
-  "js_of_ocaml-compiler" {= "6.2.0"}
   "jsonm" {= "1.0.2"}
   "jst-config" {= "v0.17.0"}
   "linol" {= "0.10"}
   "linol-lwt" {= "0.10"}
-  "logs" {= "0.9.0"}
+  "logs" {= "0.10.0"}
   "lwt" {= "5.9.2"}
-  "menhir" {= "20250912"}
-  "menhirCST" {= "20250912"}
-  "menhirLib" {= "20250912"}
-  "menhirSdk" {= "20250912"}
-  "mtime" {= "2.1.0"}
-  "num" {= "1.6"}
-  "ocaml" {= "5.3.0"}
-  "ocaml-base-compiler" {= "5.3.0"}
-  "ocaml-compiler" {= "5.3.0"}
+  "menhir" {= "20260209"}
+  "menhirCST" {= "20260209"}
+  "menhirGLR" {= "20260209"}
+  "menhirLib" {= "20260209"}
+  "menhirSdk" {= "20260209"}
+  "ocaml" {= "5.4.1"}
+  "ocaml-base-compiler" {= "5.4.1"}
+  "ocaml-compiler" {= "5.4.1"}
   "ocaml-compiler-libs" {= "v0.17.0"}
   "ocaml-config" {= "3"}
   "ocaml-options-vanilla" {= "1"}
@@ -57,10 +53,10 @@ depends: [
   "ocaml_intrinsics_kernel" {= "v0.17.1"}
   "ocamlbuild" {= "0.16.1"}
   "ocamlfind" {= "1.9.8"}
+  "ocamlgraph" {= "2.2.0"}
   "ocplib-endian" {= "1.2"}
   "octavius" {= "1.2.2"}
-  "ordering" {= "3.20.2"}
-  "parsexp" {= "v0.17.0"}
+  "ordering" {= "3.22.2"}
   "pp" {= "2.0.0"}
   "ppx_assert" {= "v0.17.0"}
   "ppx_base" {= "v0.17.0"}
@@ -79,25 +75,24 @@ depends: [
   "ppx_sexp_conv" {= "v0.17.1"}
   "ppx_yojson_conv" {= "v0.17.1"}
   "ppx_yojson_conv_lib" {= "v0.17.0"}
-  "ppxlib" {= "0.37.0"}
+  "ppxlib" {= "0.38.0"}
   "ppxlib_jane" {= "v0.17.4"}
-  "sedlex" {= "3.7"}
   "seq" {= "base"}
-  "sexplib" {= "v0.17.0"}
   "sexplib0" {= "v0.17.0"}
   "stdio" {= "v0.17.0"}
   "stdlib-shims" {= "0.3.0"}
-  "stdune" {= "3.20.2"}
+  "stdune" {= "3.22.2"}
   "stringext" {= "1.6.0"}
   "terminal_size" {= "0.2.0"}
   "time_now" {= "v0.17.0"}
   "toml" {= "7.1.0"}
-  "topkg" {= "1.1.0"}
-  "trace" {= "0.10"}
+  "top-closure" {= "3.22.2"}
+  "topkg" {= "1.1.1"}
+  "trace" {= "0.12"}
   "uri" {= "4.4.0"}
   "uutf" {= "1.0.4"}
-  "why3" {= "1.8.2"}
-  "why3find" {= "1.2.0+dev"}
+  "why3" {= "git-2c0f2992"}
+  "why3find" {= "git-eab37557"}
   "xmlm" {= "1.4.0"}
   "yojson" {= "2.2.2"}
   "zarith" {= "1.14"}
@@ -117,6 +112,12 @@ build: [
   ]
 ]
 pin-depends: [
-  ["why3.1.8.2" "git+ssh://git@gitlab.inria.fr/why3/why3#hyp_name"]
-  ["why3find.1.2.0+dev" "git+https://git.frama-c.com/pub/why3find#why3-master"]
+  [
+    "why3.git-2c0f2992"
+    "git+https://gitlab.inria.fr/why3/why3.git#2c0f2992af85f82f3eda0f158dcf10e62e0db875"
+  ]
+  [
+    "why3find.git-eab37557"
+    "git+https://git.frama-c.com/pub/why3find.git#eab37557d3e24e1913a3c4f44bc5528ef497c6c9"
+  ]
 ]

--- a/vscode/src/extension.ts
+++ b/vscode/src/extension.ts
@@ -60,6 +60,52 @@ const languages = [
   { scheme: "file", language: "why3proof" },
 ];
 
+function cargoCreusot(args: Array<string>): { status: number|null, stdout: string|Buffer, stderr: string|Buffer } {
+  const options: child_process.SpawnSyncOptions = {};
+  const home = creusotHome();
+  const dataHome = creusotDataHome();
+  const configHome = creusotConfigHome();
+  if (home !== "" || dataHome !== "" || configHome !== "") {
+    const env = { ...process.env };
+    if (home !== "") {
+      env.HOME = home;
+    }
+    if (dataHome !== "") {
+      env.XDG_DATA_HOME = dataHome;
+    }
+    if (configHome !== "") {
+      env.XDG_CONFIG_HOME = configHome;
+    }
+    options.env = env;
+  }
+  return child_process.spawnSync("cargo", args, options);
+}
+
+function getCreusotVersion(): Array<number>|null {
+  const raw_output = cargoCreusot(["creusot", "version"]);
+  if (raw_output.status != 0) {
+    window.showErrorMessage("Could not get cargo creusot version: failed to run `cargo creusot version`");
+    return null;
+  }
+  const output = typeof raw_output.stdout == 'string' ? raw_output.stdout : raw_output.stdout.toString();
+  const match = output.match(/cargo-creusot (\d+)\.(\d+)\.(\d+)(-[a-z]*)?/);
+  if (match === null) {
+    window.showErrorMessage("Could not get cargo creusot version: failed to parse output of `cargo creusot version`");
+    return null;
+  }
+  return [+match[1], +match[2], +match[3]];
+}
+
+const version = getCreusotVersion();
+
+function creusotVersionAtLeast(major: number, minor: number, fix: number): boolean {
+  if (version === null) {
+    // Either Creusot could not be found, or Creusot changed its version output; we can't do anything about the former so we assume the latter here.
+    return true
+  }
+  return major < version[0] || major == version[0] && (minor < version[1] || minor == version[1] && (fix < version[2] || fix == version[2]))
+}
+
 function startServer(context): LanguageClient {
   const outputChannel = vscode.window.createOutputChannel("Creusot IDE");
   const traceOutputChannel = vscode.window.createOutputChannel("Creusot IDE Trace");
@@ -140,25 +186,8 @@ async function createTests(client: LanguageClient) {
           // Save the rust file
           if (test.uri) { await vscode.workspace.save(test.uri); }
           process.chdir(rootPath);
-          const options: child_process.SpawnSyncOptions = {};
-          const home = creusotHome();
-          const dataHome = creusotDataHome();
-          const configHome = creusotConfigHome();
-          if (home !== "" || dataHome !== "" || configHome !== "") {
-            const env = { ...process.env };
-            if (home !== "") {
-              env.HOME = home;
-            }
-            if (dataHome !== "") {
-              env.XDG_DATA_HOME = dataHome;
-            }
-            if (configHome !== "") {
-              env.XDG_CONFIG_HOME = configHome;
-            }
-            options.env = env;
-          }
           // why3find prove
-          const output = child_process.spawnSync("cargo", ["creusot", "prove", test.id], options);
+          const output = cargoCreusot(["creusot", "prove", test.id]);
           run.appendOutput(`Finishing test ${test.label}\n\r`);
           if (output.status !== 0) {
             const logs = output.stdout.toString() + "\n" + output.stderr.toString();
@@ -184,7 +213,8 @@ async function createTests(client: LanguageClient) {
     const run = controller.createTestRun(request, "Launch IDE", false);
     run.started(test0);
     process.chdir(rootPath);
-    const output = child_process.spawnSync("cargo", ["creusot", "prove", test0.id, "-i"]);
+    // FIXME: shouldn't this have the same options as the other spawnSync above?
+    const output = cargoCreusot(["creusot", "prove", test0.id, "-i"]);
     run.end();
   }
   const debugProfile = controller.createRunProfile("Debug", vscode.TestRunProfileKind.Debug, debugHandler);

--- a/vscode/src/extension.ts
+++ b/vscode/src/extension.ts
@@ -187,7 +187,7 @@ async function createTests(client: LanguageClient) {
           if (test.uri) { await vscode.workspace.save(test.uri); }
           process.chdir(rootPath);
           // why3find prove
-          const output = cargoCreusot(["creusot", "prove", test.id]);
+          const output = cargoCreusot(creusotVersionAtLeast(0, 12, 0) ? ["creusot", test.id] : ["creusot", "prove", test.id]);
           run.appendOutput(`Finishing test ${test.label}\n\r`);
           if (output.status !== 0) {
             const logs = output.stdout.toString() + "\n" + output.stderr.toString();
@@ -213,8 +213,7 @@ async function createTests(client: LanguageClient) {
     const run = controller.createTestRun(request, "Launch IDE", false);
     run.started(test0);
     process.chdir(rootPath);
-    // FIXME: shouldn't this have the same options as the other spawnSync above?
-    const output = cargoCreusot(["creusot", "prove", test0.id, "-i"]);
+    const output = cargoCreusot(creusotVersionAtLeast(0, 12, 0) ? ["creusot", test0.id, "-i"] : ["creusot", "prove", test0.id, "-i"]);
     run.end();
   }
   const debugProfile = controller.createRunProfile("Debug", vscode.TestRunProfileKind.Debug, debugHandler);
@@ -239,7 +238,7 @@ export function activate(context: ExtensionContext) {
     vscode.TaskScope.Workspace,
     "Creusot Build",
     "creusot",
-    new vscode.ShellExecution("cargo", ["creusot"])
+    new vscode.ShellExecution("cargo", creusotVersionAtLeast(0, 12, 0) ? ["creusot"] : ["creusot", "--only=coma"])
   )
   registerCommand(context, "creusot.build", async () => {
     const exec = await vscode.tasks.executeTask(creusotBuild)
@@ -267,9 +266,9 @@ export function activate(context: ExtensionContext) {
       command: "why3find",
     },
     vscode.TaskScope.Workspace,
-    "Why3find Prove Everything",
+    "Creusot Prove",
     "creusot",
-    new vscode.ShellExecution("cargo", ["creusot", "prove"])
+    new vscode.ShellExecution("cargo", creusotVersionAtLeast(0, 12, 0) ? ["creusot"] : ["creusot", "prove"])
   )
   registerCommand(context, "creusot.prove", async () => {
     const exec = await vscode.tasks.executeTask(creusotProve)

